apps/api/src/lib/flat-parse.ts

```diff
@@ -1,118 +1,52 @@
- *   email: z.string().email("Invalid email"),
- *   age: z.number().min(18, "Must be 18 or older")
- * });
- *
- * const result = flatParse(userSchema, { name: "", email: "invalid", age: 16 });
- * if (!result.success) {
- *   console.log(result.error);
- *   // Output: "Name is required, Invalid email, Must be 18 or older"
- * }
- * ```
- *
- * @example
- * ```typescript
- * // In an Express route handler
- * app.post('/users', (req, res) => {
- *   const result = flatParse(UserSchema, req.body);
- *   if (!result.success) {
- *     return res.status(400).json({ error: result.error });
- *   }
- *
- *   // result.data is fully typed and validated
- *   const user = result.data;
- *   // ... process user
- * });
- * ```
- *
- * @example
- * ```typescript
- * // Comparison with Zod's safeParse
- * // Zod's approach:
- * const zodResult = userSchema.safeParse(input);
- * if (!zodResult.success) {
- *   // zodResult.error.issues is an array of detailed error objects
- *   const errors = zodResult.error.issues.map(issue =>
- *     `${issue.path.join('.')}: ${issue.message}`
- *   );
- *   console.log(errors);
- * }
- *
- * // This utility's approach:
- * const flatResult = flatParse(userSchema, input);
- * if (!flatResult.success) {
- *   // flatResult.error is a single string with all messages
- *   console.log(flatResult.error);
- * }
- * ```
+ * @param requestPart - Optional string to specify which part of the request is being validated (for error messages)
+ * @returns The validated data, fully typed as z.TypeOf<TSchema>
+ * @throws {ApiError} If validation fails, throws with a concatenated error message
  */
-function flatParse<TSchema extends z.ZodType>(
+export function flatParse<TSchema extends z.ZodType>(
   schema: TSchema,
   input: unknown,
-): FlatParseResult<z.TypeOf<TSchema>> {
-  const { success, error, data } = z.safeParse(schema, input);
-  if (success) {
-    return { success: true, data: data };
+  requestPart?: string,
+): z.infer<TSchema> {
+  const result = schema.safeParse(input);
+  if (result.success) {
+    return result.data;
   } else {
-    const message = error?.issues.map((i) => i.message).join(", ");
-    return { success: false, error: message };
+    const message = result.error.issues
+      .map((issue) => {
+        // Clean up the error message by removing redundant "Invalid input: " prefix, include the
+        // invalid field (if applicable), and improve casing for mid-sentence words.
+        const cleanMessage = issue.message
+          .replace(/^Invalid input: /, "")
+          .replace(/^./, (char) => char.toLowerCase());
+        const message = (fieldPath?: string) =>
+          `${fieldPath ? `${fieldPath} (${cleanMessage})` : cleanMessage}`;
+        return issue.path.length > 0
+          ? message(issue.path.join("."))
+          : message();
+      })
+      .join("; ");
+    if (requestPart) {
+      throw new ApiError(400, `Invalid request ${requestPart}: ${message}`);
```

Comment:
if you pass some context like `body` or `query` to the method, it'll make the error message more descriptive like `Invalid request body` instead of purely `Invalid request`.

note: this is not actually used in this PR within the admin controller, but it could be helpful